File: Chapter 3/Section 3.2/butter.py
"""
ID: hsfncd31
LANG: PYTHON3
TASK: butter
TLE at Case#9
"""
import os
import typing
import collections
import heapq


def dijkstra(graph: typing.DefaultDict[int, typing.List[typing.Tuple[int, int]]], num_vertices: int, source: int)\
        -> typing.List[typing.Union[int, None]]:
    processed = set()
    distance = [None] * num_vertices
    distance[source] = 0
    # (distance, vertex), may have duplicate vertex
    nearest_vertices = [(0, source)]
    while len(nearest_vertices):
        _, v = heapq.heappop(nearest_vertices)
        if v in processed:
            continue
        processed.add(v)
        for v2, l in graph[v]:
            if distance[v2] is None or distance[v] + l < distance[v2]:
                distance[v2] = distance[v] + l
                heapq.heappush(nearest_vertices, (distance[v2], v2))
    return distance


def main():
    base_filename = 'test' if os.name == 'nt' else 'butter'
    with open('{}.in'.format(base_filename), 'r') as infile,\
            open('{}.out'.format(base_filename), 'w') as outfile:  # type: typing.IO[str]
        def get_line() -> str:
            return infile.readline().rstrip('\n')

        def out_print(*args, **kwargs) -> None:
            kwargs['file'] = outfile
            print(*args, **kwargs)

        num_cows, num_pastures, num_paths = map(int, get_line().split())
        num_cows_in_pasture = collections.defaultdict(int)
        for _ in range(num_cows):
            num_cows_in_pasture[int(get_line()) - 1] += 1
        # adjacent linked list
        # { v1: [(v2, length)] }
        graph = collections.defaultdict(list)
        for _ in range(num_paths):
            a, b, l = map(int, get_line().split())
            if a == b:
                continue
            a -= 1
            b -= 1
            graph[a].append((b, l))
            graph[b].append((a, l))

        answer = -1  # I'd like to use None but PyCharm gives warning
        for sugar_pasture in range(num_pastures):
            distance = dijkstra(graph, num_pastures, sugar_pasture)
            total_cost = 0
            for pasture, num_cows_grazing in num_cows_in_pasture.items():
                if distance[pasture] is None:
                    total_cost = -1
                    break
                total_cost += distance[pasture] * num_cows_grazing
            if total_cost != -1 and (answer == -1 or total_cost < answer):
                answer = total_cost

        out_print(answer)


if __name__ == '__main__':
    main()

File: Chapter 3/Section 3.2/butter.cpp
/*
ID: hsfncd31
TASK: butter
LANG: C++                 
*/
#include <fstream>
#include <map>
#include <vector>
#include <queue>

std::vector<int> dijkstra(
        std::map<int, std::vector<std::pair<int, int>>> graph,
        int num_vertices, int source) {
    std::vector<int> distance(num_vertices, -1);
    distance[source] = 0;
    std::vector<bool> processed(num_vertices);
    // [(-distance, vertex)]
    std::priority_queue<std::pair<int, int>> q;
    q.push(std::make_pair(0, source));
    while (!q.empty()) {
        int v = q.top().second;
        q.pop();
        if (processed[v])
            continue;
        processed[v] = true;
        for (auto &kv: graph[v]) {
            int v2 = kv.first, l = kv.second;
            if (distance[v2] == -1 || distance[v] + l < distance[v2]) {
                distance[v2] = distance[v] + l;
                q.push(std::make_pair(-distance[v2], v2));
            }
        }
    }
    return distance;
}

int main() {
    std::fstream in("butter.in", std::fstream::in), out("butter.out", std::fstream::out);

    int num_cows, num_pastures, num_paths;
    in >> num_cows >> num_pastures >> num_paths;
    std::map<int, int> num_cows_in_pasture;
    for (int _ = 0; _ < num_cows; ++_) {
        int pasture;
        in >> pasture;
        num_cows_in_pasture[pasture - 1] += 1;
    }
    // { v: [(v2, l)] }
    std::map<int, std::vector<std::pair<int, int>>> graph;
    for (int _ = 0; _ < num_paths; ++_) {
        int a, b, l;
        in >> a >> b >> l;
        if (a == b)
            continue;
        a -= 1;
        b -= 1;
        graph[a].push_back(std::make_pair(b, l));
        graph[b].push_back(std::make_pair(a, l));
    }

    int answer = -1;
    for (int sugar_pasture = 0; sugar_pasture < num_pastures; ++sugar_pasture) {
        std::vector<int> distance = dijkstra(graph, num_pastures, sugar_pasture);
        int total_cost = 0;
        for (auto &kv: num_cows_in_pasture) {
            if (distance[kv.first] == -1) {
                total_cost = -1;
                break;
            }
            total_cost += distance[kv.first] * kv.second;
        }
        if (total_cost != -1 && (answer == -1 || total_cost < answer))
            answer = total_cost;
    }

    out << answer << '\n';

    in.close();
    out.close();
}
